aips/aip-54.md

```diff
@@ -0,0 +1,107 @@
+2. Provide a powerful object-based framework for managing code deployment and upgrading.
+3. Reduce storage footprint of code deployment
+
+## Existing solutions
+Current existing solutions:
+1. Deploy code directly to an account the developer owns. This has two main drawbacks:
+* During development, if the developer needs to make incompatible changes, they would need to throw away the account and deploy to a different one to avoid naming conflicts. Alternatively, they would need to change their package and module names.
+* Deploying directly to an account means that account has full control over upgrading the code and cannot give this permission away, e.g. to a governance entity. This limits programmatic control, flexibility and decentralization of protocols deployed on Aptos.
+2. Deploy code to a newly created resource account. This is similar to object deployment but suffers from:
+* Similar issue with incompatible upgrades during development. Developer would need to play around with seeds to generate a different resource account, which is confusing.
+* Storage inefficiency - creating a resource account to host the code means at least 2 resources created (especially when the Account resource is not needed).
+* Hard to use upgrade process. Upgrade permission is controlled via the resource account's SignerCapability which is confusing to claim and use to upgrade code.
+
+## Specification
+The new object deployment flow will contain 3 specific components:
+1. A new object_code module that offers the different APIs for deploying, upgrading, etc. using objects
+2. A new resource PackageRegistryObject that wraps the existing PackageRegistry resource into the object resource group for storage efficiency.
+3. Improvements to the Aptos CLI to simplify code deployment and upgrade flows.
+
+In addition, the following improvements can be considered:
+1. Extra metadata in PackageRegistry to support adding custom metadata such as git commit of the latest code version.
+2. Allow specifying the current code object address in Move.toml to make upgrading easier to manage.
+3. Allowing PackageMetadata as a struct input to the entry function. This would avoid having to rely on the CLI to generate the bytes for the PackageMetadata which is confusing and easy to make mistakes.
+4. Ability to incrementally upload part of the code and publish the final package. This allows deploying packages larger than the 64KB size limit per transaction.
+5. VM changes: The ability to deploy code without specifying the address of the code in advance. This can be used in conjunction with dynamically creating a new object to host the newly deployed code.
+
+### New object_code module
+```rust
+module aptos_framework::object_code {
+  #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+  struct PublishRef has key {
+    inner: Object<PackageRegistryObject>,
+    // Extend ref of the code object.
+    extend_ref: ExtendRef,
+  }
```

Comment:
This is an object itself besides `PackageRegistryObject`?
My understanding is:
depending on whether the contract is upgradable, `extend_ref` should be an optional resource in `PackageRegistryObject`. while `PublisherRef` is unnecessary since object ownership can be used to do authorization for getting the `extend_ref`.

Reply:
Object ownership could work too. We still need to keep the ExtendRef for that case though